depfile: reduce dag with `--use-buildcache` flag.

haampie opened the pull request:
When installing some/all specs from a buildcache, build edges are pruned
from those specs. This can result in a much smaller effective DAG. Until
now, `spack env depfile` would always generate a full DAG.

Ths PR adds the `spack env depfile --use-buildcache` flag that was
introduced for `spack install` before. This way, not only can we drop
build edges, but also we can automatically set the right buildcache
related flags on the specific specs that are gonna get installed.

On top of that, this PR introduces `spack env depfile [spec 1] [spec 2]...`
that reduces the dag to particular specs of interest in the environment.

This way we get parallel installs of binary deps of a particular spec
that's built from sources, without redundancy, which is useful for
Gitlab CI.

Example use:

```console
# install from buildcache w/o build deps
spack env depfile -o Makefile --use-buildcache only && make -j8

# drop build deps of deps
spack env depfile -o Makefile --use-buildcache dependencies:only && make -j8

# restrict environment install to pkg-a and pkg-b + deps
# build those two from sources
# dependencies are pulled from a buildcache, their build deps dropped
spack env depfile -o Makefile --use-buildcache package:never,dependencies:only pkg-a pkg-b && make -j8 
```

haampie commented:
@spackbot fix style

haampie commented:
@alalazo mentioned:

```
spack env depfile --[build-|install-]from-sources=all/root/none
```

If we do that, we can immediately add the right `--cache-only` type of flags too.

scheibelp commented on lib/spack/spack/cmd/env.py:
```diff
@@ -613,17 +620,103 @@ def env_depfile_setup_parser(subparser):
+    pass
+
+
+def traverse_breadth_first(visitor, specs=[]):
```
`Spec.traverse(order='pre')` should match this in all the desired ways and might avoid some corner cases like:

```
1____
| \  \
2  3  4
|  |  |
|  5  6
|_/__/
|
7
```

i.e., we want something more-sophisticated than breadth first or we may end up visiting `7` before `[5, 6]` (if your implementation avoids this, we should probably call it something like "topological" rather than "breadth-first").

haampie replied:
Hm, the point is I have multiple root specs, and need to handle the roots differently from the dependencies. Given that the second root may appear as a dependency of the first root, I don't want depth-first search. Breadth-first search initialized with all root specs guarantees that roots are found at depth 0 and that all connected nodes are visited once. I don't see the need for topological sort, for make targets all I need is the adjacency list.

See the bit

```python
    def neighbors(self, node):
        # At depth zero we return build / test deps too
        if node.depth == 0:
            return node.spec.dependencies(deptype="all")
        # Otherwise we just return link/run type deps
        return node.spec.dependencies(deptype=("link", "run"))
```

haampie committed:
depfile: buildcache support

When installing some/all specs from a buildcache, build edges are pruned
from those specs. This can result in a much smaller effective DAG. Until
now, `spack env depfile` would always generate a full DAG.

Ths PR adds the `spack env depfile --use-buildcache` flag that was
introduced for `spack install` before. This way, not only can we drop
build edges, but also we can automatically set the right buildcache
related flags on the specific specs that are gonna get installed.

This way we get parallel installs of binary deps without redundancy,
which is useful for Gitlab CI.

haampie commented:
This is ready for review, again. Includes tests now.

haampie commented:
@scheibelp I've moved the breadth-first traversal bits to #33406 so that this PR doesn't introduce its own "custom" traversal.
